src/kOS.Safe/Compilation/KS/Compiler.cs

```diff
@@ -413,76 +413,85 @@ private void PreProcessChildNodes(ParseNode node)
         private void PreProcessOnStatement(ParseNode node)
         {
             NodeStartHousekeeping(node);
+            nextBraceIsFunction = true; // triggers aren't really functions but act like it a lot.
+
             int expressionHash = ConcatenateNodes(node).GetHashCode();
             string triggerIdentifier = "on-" + expressionHash.ToString();
             Trigger triggerObject = context.Triggers.GetTrigger(triggerIdentifier);
-            triggerObject.SetTriggerVariable(GetIdentifierText(node));
 
             currentCodeSection = triggerObject.Code;
-            AddOpcode(new OpcodePush(triggerObject.VariableNameOldValue));
-            AddOpcode(new OpcodePush(triggerObject.VariableName));
+            // Put the old value on top of the stack for equals comparison later:
+            AddOpcode(new OpcodePush(triggerObject.OldValueIdentifier));
+            AddOpcode(new OpcodeEval());
+            // eval the expression for the new value, and leave it on the stack twice.
+            VisitNode(node.Nodes[1]);
+            AddOpcode(new OpcodeEval());
+            AddOpcode(new OpcodeDup());
+            // Put one of those two copies of the new value into the old value identifier for next time:
+            AddOpcode(new OpcodePush(triggerObject.OldValueIdentifier));
+            AddOpcode(new OpcodeSwap());
+            AddOpcode(new OpcodeStoreGlobal());
+            // Use the other dup'ed copy of the new value to actually do the equals
+            // comparison with the old value that's still under it on the stack:
             AddOpcode(new OpcodeCompareEqual());
-            AddOpcode(new OpcodeLogicNot());
-            Opcode branchOpcode = AddOpcode(new OpcodeBranchIfFalse());
+            OpcodeBranchIfFalse branchToBody = new OpcodeBranchIfFalse();
+            branchToBody.Distance = 3;
+            AddOpcode(branchToBody);
+            AddOpcode(new OpcodePush(true));       // wasn't triggered yet, so preserve.
+            AddOpcode(new OpcodeReturn((short)0)); // premature return because it wasn't triggered
```

Comment:
The edits from lines 423 to 441 (I'm using the new (green) numbering when I say that) are all because of the change to make ``on`` triggers work on any arbitrary expression, rather than requiring just an identifier variable.  Now it has to run the expression, eval the result (thus de-referencing it if it *is* just a variable name), and use the result of THAT as the value to compare against the previous stored value.

Possible weird side effect?  If the result of the expression is a reference to a thing (i.e. a Vessel or a Body or something like that), then as long as the ``on`` trigger is still active, that thing won't orphan and get garbage collected.